File: AIReader/AIReader/Services/AI/ChatAPI.swift
import Foundation

/// Chat completions and the model list, against any OpenAI-compatible service.
/// Request bodies are assembled as plain JSON so tool schemas stay readable.
enum ChatAPI {
    enum APIError: LocalizedError {
        case invalidEndpoint(String)
        case rateLimited
        case http(status: Int, body: String)
        case noChoices

        var errorDescription: String? {
            switch self {
            case let .invalidEndpoint(endpoint): "“\(endpoint)” is not a valid endpoint URL."
            case .rateLimited: "The service is rate limiting this token. Try again shortly."
            case let .http(status, body): "The request failed (\(status)): \(body)"
            case .noChoices: "The service returned no answer."
            }
        }
    }

    static func chat(
        settings: AISettings,
        messages: [ChatMessage],
        tools: [[String: Any]] = [],
        jsonMode: Bool = false
    ) async throws -> ChatMessage {
        guard !settings.usesMock else { return MockAI.reply(to: messages) }
        guard let url = settings.chatURL else {
            throw APIError.invalidEndpoint(settings.endpoint)
        }

        let signature = "\(settings.endpoint)|\(settings.model)"
        var quirks = await RequestQuirkStore.shared.quirks(for: signature)

        // A service that rejects a parameter says which one, so drop or rename
        // it and try again rather than failing the lookup.
        for _ in 0...RequestQuirk.allCases.count {
            var request = URLRequest(url: url)
            request.httpMethod = "POST"
            request.timeoutInterval = 45
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.httpBody = try JSONSerialization.data(
                withJSONObject: body(
                    settings: settings,
                    messages: messages,
                    tools: tools,
                    jsonMode: jsonMode,
                    quirks: quirks
                )
            )

            do {
                let data = try await send(request, apiKey: settings.apiKey)
                guard let message = try JSONDecoder().decode(Completion.self, from: data)
                    .choices.first?.message
                else { throw APIError.noChoices }
                return message
            } catch let APIError.http(status, body) where status == 400 {
                guard let quirk = RequestQuirk.named(inErrorBody: body),
                      quirks.insert(quirk).inserted
                else { throw APIError.http(status: status, body: body) }
                await RequestQuirkStore.shared.learn(quirk, for: signature)
            }
        }

        throw APIError.noChoices
    }

    static func body(
        settings: AISettings,
        messages: [ChatMessage],
        tools: [[String: Any]],
        jsonMode: Bool,
        quirks: Set<RequestQuirk>
    ) throws -> [String: Any] {
        var body: [String: Any] = [
            "model": settings.model,
            "messages": try JSONSerialization.jsonObject(with: JSONEncoder().encode(messages))
        ]
        body[quirks.contains(.completionTokens) ? "max_completion_tokens" : "max_tokens"] = 700
        if !quirks.contains(.defaultTemperature) {
            body["temperature"] = 0.2
        }
        if quirks.contains(.noReasoning) {
            body["reasoning_effort"] = "none"
        }
        if !tools.isEmpty {
            body["tools"] = tools
            body["tool_choice"] = "auto"
        }
        if jsonMode {
            body["response_format"] = ["type": "json_object"]
        }
        return body
    }

    /// The model names the endpoint offers, sorted.
    static func models(settings: AISettings) async throws -> [String] {
        guard !settings.usesMock else { return MockAI.models }
        guard let url = settings.modelsURL else {
            throw APIError.invalidEndpoint(settings.endpoint)
        }

        var request = URLRequest(url: url)
        request.timeoutInterval = 20
        request.setValue("application/json", forHTTPHeaderField: "Accept")

        let data = try await send(request, apiKey: settings.apiKey)
        let list = try JSONDecoder().decode(ModelList.self, from: data)
        return Set(list.data.map(\.id)).sorted()
    }

    private static func send(_ request: URLRequest, apiKey: String) async throws -> Data {
        var request = request
        if !apiKey.isEmpty {
            request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        }

        let (data, response) = try await URLSession.shared.data(for: request)
        if let http = response as? HTTPURLResponse, !(200...299).contains(http.statusCode) {
            if http.statusCode == 429 { throw APIError.rateLimited }
            throw APIError.http(
                status: http.statusCode,
                body: String(data: data, encoding: .utf8) ?? ""
            )
        }
        return data
    }

    private struct Completion: Decodable {
        struct Choice: Decodable { let message: ChatMessage }
        let choices: [Choice]
    }

    private struct ModelList: Decodable {
        struct Model: Decodable { let id: String }
        let data: [Model]
    }
}

File: AIReaderKindle/src/Services/ChatApi.cpp
#include "ChatApi.hpp"

#include "../Domain/AI/MockAI.hpp"
#include "../Domain/AI/RequestQuirks.hpp"
#include "Http.hpp"

#include <algorithm>
#include <set>

namespace ChatApi {

namespace {

using Http::Response;
using Http::send;

Json requestBody(
    const AiSettings& settings,
    const std::vector<ChatMessage>& messages,
    const Json& tools,
    bool jsonMode,
    const std::set<RequestQuirk>& quirks)
{
    Json body = Json::object();
    body.set("model", settings.model);
    Json list = Json::array();
    for (const auto& message : messages) list.push(message.toJson());
    body.set("messages", list);
    body.set(quirks.count(RequestQuirk::CompletionTokens) ? "max_completion_tokens" : "max_tokens", 700);
    if (!quirks.count(RequestQuirk::DefaultTemperature)) body.set("temperature", 0.2);
    if (quirks.count(RequestQuirk::NoReasoning)) body.set("reasoning_effort", "none");
    if (tools.size() > 0) {
        body.set("tools", tools);
        body.set("tool_choice", "auto");
    }
    if (jsonMode) {
        Json format = Json::object();
        format.set("type", "json_object");
        body.set("response_format", format);
    }
    return body;
}

}  // namespace

void initialize() {
    Http::initialize();
}

ChatMessage chat(const AiSettings& settings, const std::vector<ChatMessage>& messages, const Json& tools, bool jsonMode) {
    if (settings.usesMock()) return MockAI::reply(messages);
    std::string url = settings.chatUrl();
    if (url.find("://") == std::string::npos) throw Error("“" + settings.endpoint + "” is not a valid endpoint URL.");

    std::string signature = settings.endpoint + "|" + settings.model;
    std::set<RequestQuirk> quirks = RequestQuirkStore::shared().quirks(signature);

    // A service that rejects a parameter says which one, so drop or rename it
    // and try again rather than failing the lookup.
    for (int attempt = 0; attempt <= requestQuirkCount; ++attempt) {
        std::string body = requestBody(settings, messages, tools, jsonMode, quirks).dump();
        Response response = send(url, settings.token(), &body, 45);

        if (response.status == 400) {
            auto quirk = requestQuirkNamed(response.body);
            if (quirk && quirks.insert(*quirk).second) {
                RequestQuirkStore::shared().learn(*quirk, signature);
                continue;
            }
        }
        if (response.status < 200 || response.status >= 300) {
            throw Error("The request failed (" + std::to_string(response.status) + "): " + response.body);
        }
        auto json = Json::parse(response.body);
        if (!json || json->at("choices").size() == 0) throw Error("The service returned no answer.");
        return ChatMessage::fromJson(json->at("choices").at(0).at("message"));
    }
    throw Error("The service returned no answer.");
}

std::vector<std::string> models(const AiSettings& settings) {
    if (settings.usesMock()) return MockAI::models;
    std::string url = settings.modelsUrl();
    if (url.find("://") == std::string::npos) throw Error("“" + settings.endpoint + "” is not a valid endpoint URL.");

    Response response = send(url, settings.token(), nullptr, 20);
    if (response.status < 200 || response.status >= 300) {
        throw Error("The request failed (" + std::to_string(response.status) + "): " + response.body);
    }
    auto json = Json::parse(response.body);
    if (!json) throw Error("The model list could not be read.");
    std::set<std::string> names;
    for (const auto& model : json->at("data").items()) {
        if (model.at("id").isString()) names.insert(model.at("id").string());
    }
    return std::vector<std::string>(names.begin(), names.end());
}

}  // namespace ChatApi
